### PR TITLE
feat: implement `ContractFactory` for contract deployment

### PR DESCRIPTION
**Pull request by xJonathanLEI**
Resolves task 3 of #6:

> add type for deploying artifact through `starknet_providers::Provider`

### CODE REVIEW

**Comment by xJonathanLEI** [comment]
Note that this PR adds an integration test for deploying a contract to `alpha-goerli`. To avoid unnecessarily spamming the network, when developing locally, you're advised to skip running the test case unless it's relevant.

File: starknet-contract/src/factory.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        rng.fill_bytes(&mut salt_buffer[1..]);
+
+        self.provider
+            .add_transaction(
+                TransactionRequest::Deploy(DeployTransactionRequest {
+                    contract_address_salt: H256::from(salt_buffer),
+                    contract_definition: ContractDefinition {
+                        program: self.compressed_program.clone(),
+                        entry_points_by_type: self.entry_points_by_type.clone(),
+                        abi: Some(self.abi.clone()),
```
Comment (milancermak) [inline starknet-contract/src/factory.rs:72]: Could there actually be a contract without an ABI?

**Review by milancermak — APPROVED** [review]
All good

File: starknet-contract/src/factory.rs (hunk 1 above)
Comment (xJonathanLEI) [reply starknet-contract/src/factory.rs:72]: Yes. Just manually change this to `None` and it'll still work.

Also as noted [here](https://community.starknet.io/t/remove-abi-from-contract-deployment-request-and-get-code-response/308/2), there's really no way for the sequencer to validate the ABI. I think that's why they made it optional.

I was thinking about whether we should give users a choice on this too, by maybe making `abi` optional on the artifact type. What do you think?

Comment (xJonathanLEI) [reply starknet-contract/src/factory.rs:72]: So maybe we should expose the compression logic on the `Program` type, and have users call the provider directly if they want to skip ABI?

Actually, I was initially going to build the compression logic there. Don't remember why I ended up putting in the factory ctor, maybe too sleepy lol.

**Comment by xJonathanLEI** [comment]
Anyways. Lemme merge this first and refactor later on if needed :)

File: starknet-contract/src/factory.rs (hunk 1 above)
Comment (milancermak) [reply starknet-contract/src/factory.rs:72]: I have no opinion on this, but a cool hack on deploying the fake ABI 🎩

**Pull request merged**